File: 22-generate-parentheses/generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:

        def backtrack(open_len, close_len, current):

            if len(current) == 2 * n:
                result.append(current)

            if open_len > close_len:
                backtrack(open_len, close_len + 1, current + ")")
            
            if open_len < n:
                backtrack(open_len + 1, close_len, current + "(")


        result = []
        backtrack(0, 0, "")
        return result

File: 22-generate-parentheses/generate-parentheses.java
class Solution {

    public void backtrack(List<String> result, int openCount, int closeCount, int n, String current) {

        if (current.length() == 2 * n) {
            result.add(current);
        }

        if (openCount > closeCount) {
            backtrack(result, openCount, closeCount + 1, n, current + ")");
        }
        if (openCount < n) {
            backtrack(result, openCount + 1, closeCount, n, current + "(");
        }
    }

    public List<String> generateParenthesis(int n) {

        List<String> result = new ArrayList<>();
        backtrack(result, 0, 0, n, "");

        return result;
    }
}